File: core/src/main/scala/tripaint/coords/GlobalPixCoords.scala
package tripaint.coords

/** Like TriImageCoords but for pixels (on the entire area) */
case class GlobalPixCoords(value: Long) extends AnyVal {
  inline def x = (value >> 32).toInt
  inline def y = (value & 0xffffffff).toInt

  infix def distanceSq(other: GlobalPixCoords): Double = {
    val dx = other.x - x
    val dy = other.y - y
    val otherY = other.y + (if other.x % 2 == 0 then 1.0 / 3 else 2.0 / 3)
    val thisY = y + (if x % 2 == 0 then 1.0 / 3 else 2.0 / 3)
    val actualDy = otherY - thisY
    val xx = dx * 0.5 + dy * 0.5
    // It's like magic!
    val yy = actualDy * GlobalPixCoords.sqrt3 / 2
    xx * xx + yy * yy
  }

  def neighbours: Seq[GlobalPixCoords] = {
    for (xx, yy) <- Seq(
        (x - 1, y),
        if x % 2 == 0 then (x + 1, y - 1) else (x - 1, y + 1),
        (x + 1, y)
      )
    yield GlobalPixCoords(xx, yy)
  }

  def cell: GlobalPixCoords = {
    val cy = y >> 1
    val cx = x >> 2
    val cz = (x + 1 + (y << 1)) >> 2

    if cx % 2 == cz % 2 then {
      GlobalPixCoords(cx << 1, cy)
    } else {
      GlobalPixCoords((cx << 1) + 1, cy)
    }
  }
}

object GlobalPixCoords {
  private val sqrt3: Double = math.sqrt(3)

  inline def apply(x: Int, y: Int): GlobalPixCoords = new GlobalPixCoords(
    x.toLong << 32 | (y.toLong & 0xffffffffL)
  )
}

File: core/src/main/kotlin/tripaint/coords/GlobalPixCoords.kt
package tripaint.coords

import kotlin.math.sqrt

/** Like TriImageCoords but for pixels (on the entire area) */
data class GlobalPixCoords(val value: Long) {
    val x: Int
        get() = (value shr 32).toInt()
    val y: Int
        get() = (value and 0xffffffff).toInt()

    infix fun distanceSq(other: GlobalPixCoords): Double {
        val dx = other.x - x
        val dy = other.y - y
        val otherY = other.y + (if (other.x % 2 == 0) 1.0 / 3 else 2.0 / 3)
        val thisY = y + (if (x % 2 == 0) 1.0 / 3 else 2.0 / 3)
        val actualDy = otherY - thisY
        val xx = dx * 0.5 + dy * 0.5
        // It's like magic!
        val yy = actualDy * sqrt3 / 2
        return xx * xx + yy * yy
    }

    fun neighbours(): List<GlobalPixCoords> {
        return listOf(
            Pair(x - 1, y),
            if (x % 2 == 0) Pair(x + 1, y - 1) else Pair(x - 1, y + 1),
            Pair(x + 1, y)
        ).map {
            val (xx, yy) = it
            from(xx, yy)
        }
    }

    fun cell(): GlobalPixCoords {
        val cy = y shr 1
        val cx = x shr 2
        val cz = (x + 1 + (y shl 1)) shr 2

        return if (cx % 2 == cz % 2) {
            from(cx shl 1, cy)
        } else {
            from((cx shl 1) + 1, cy)
        }
    }

    companion object {
        private val sqrt3: Double = sqrt(3.0)

        fun from(x: Int, y: Int): GlobalPixCoords = GlobalPixCoords((x.toLong() shl 32) or (y.toLong() and 0xffffffffL))
    }

}